quickcomm/models.py

```diff
@@ -1,3 +1,125 @@
+from django.contrib.auth.models import User
+from django.core.validators import URLValidator
+
+# NOTE: The models in this file do not take into account how the site will
+# interact with other APIs. Ideally, we should be able to reuse the model
+# classes, but this logic is to be implemented later.
+
+# Type fields are left out as they can be added later and are redundant. Host
+# fields are kept as the if these classes are used to represent remote authors,
+# the host field is necessary. ID is also left out.
+
+# We used UUIDs for pkeys to be more secure.
+# TODO what are the constraints on fields being null?
+
+class Author(models.Model):
+    """An author is a person associated with a user account via a one-to-one
+    relationship."""
+
+    # Example code for a one-to-one relationship taken from
+    # https://simpleisbetterthancomplex.com/tutorial/2016/07/22/how-to-extend-django-user-model.html#onetoone
+    # Author: Vitor Freitas
+    # Site: Simple is Better Than Complex
+    # Taken on: Feb 6, 2023
+
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
+    user = models.OneToOneField(User, on_delete=models.CASCADE)
+    host = models.URLField(validators=[URLValidator])
+    display_name = models.CharField(max_length=100)
+    github = models.URLField(blank=True, null=True, validators=[URLValidator])
+    profile_image = models.URLField(blank=True, null=True, validators=[URLValidator])
+
+    # TODO determine if admins are authors
+    is_admin = models.BooleanField(default=False)
+
+    def follows(self, author):
+        """Returns true if this author (self) follows the given author."""
+        return Follow.objects.filter(follower=self, following=author).exists()
+
+    def following(self, author):
+        """Returns true if this author (self) is followed by the given author."""
+        return Follow.objects.filter(follower=author, following=self).exists()
+
+
+    def __str__(self):
+        return f"{self.display_name} ({self.user.username})"
+
+class Follow(models.Model):
+    """A follow is is a many-to-many relationship between authors representing a
+    follow."""
+
+    follower = models.ForeignKey(Author, on_delete=models.CASCADE, related_name='follower')
+    following = models.ForeignKey(Author, on_delete=models.CASCADE, related_name='following')
+
+    def is_bidirectional(self):
+        """Returns true if the follow is bidirectional."""
+        return self.following.follows(self.follower)
+
+    def __str__(self):
+        return f"{self.follower.__str__()} follows {self.following.__str__()}"
+
+class Post(models.Model):
+    """A post is a post made by an author."""
+
+    class PostType(models.TextChoices):
+        TEXT = 'text/plain'
+        MD = 'text/markdown'
+        PNG = 'image/png;base64'
+        JPG = 'image/jpeg;base64'
+        APP = 'application/base64'
+
+    class PostVisibility(models.TextChoices):
+        PUBLIC = 'PUBLIC'
+        FRIENDS = 'FRIENDS'
```

Comment:
I think there should also be an 'UNLISTED' option based on this user story: As an author, I want to be able to make posts that are unlisted, that are publicly shareable by URI alone (or for embedding images).

Reply:
There is a separate model variable for unlisted in the sample API call, so it seems like both public and friends can be both unlisted and listed.

I think in that user story's case, it should be an unlisted public item? We should probably get some clarification on it though.

Reply:
I agree, that makes sense. I think for an unlisted post we just allow anyone to view it if they have the URI and it's public and we enforce someone to be logged in and a friend to see an unlisted friend post.